### ISSUE TITLE
How to speed up PPOTrainer .generate()?

### ISSUE THREAD

**Issue by RylanSchaeffer** [huggingface/trl#1900]
During PPO, I'm finding that `.generate()` is extremely slow. The following call takes ~3 and a half minutes for batch size of 64 with a 1.4B parameter policy LM:

```
ppo_trainer.generate(
                input_token_ids_list,
                pad_token_id=policy_model_tokenizer.eos_token_id,
                return_prompt=False,
                **generation_config_dict,
            )
```

How can I accelerate sampling? The same function call with `vllm` takes <30s for setup and execution, so I feel like I am doing something suboptimally.

**Comment by RylanSchaeffer** [comment]
My GPU-Util (for A100s) hovers around 35-39%. This seems very low. Maybe I'm doing something wrong?

**Comment by qgallouedec** [comment]
What version of TRL are you using?
Do you use PPOv2?

**Comment by RylanSchaeffer** [comment]
I'm using TRL 0.9.6 and PPOTrainer, not PPOv2Trainer.

**Comment by zhuyuzy** [comment]
> I'm using TRL 0.9.6 and `PPOTrainer`, not `PPOv2Trainer`.

The generation process itself is quite time-consuming, especially when using `DeepSpeed Zero-3`. You can mitigate the slowdown caused by Zero-3 by collecting the model parameters before generation and then redistributing them to the GPUs after the generation is complete. This approach helps counteract the negative effects on speed. This functionality has already been implemented in `PPOv2Trainer`, so you might want to give it a try.

**Comment by RylanSchaeffer** [comment]
I'm not using DeepSpeed. Is this advice still applicable?

**Comment by zhuyuzy** [comment]
> I'm not using DeepSpeed. Is this advice still applicable?
> 
> Follow up question: should I be using DeepSpeed?

If DeepSpeed wasn't used, then I'm not sure. 

I've only used DeepSpeed, and it works well.

**Comment by RylanSchaeffer** [comment]
Can you tell me more about (1) your code & compute setups and (2) what sort of speed you're able to attain? It might make sense for me to switch to your setup.

**Comment by RylanSchaeffer** [comment]
@zhuyuzy bump!

**Comment by zhuyuzy** [comment]
> Can you tell me more about (1) your code & compute setups and (2) what sort of speed you're able to attain? It might make sense for me to switch to your setup.

I use exactly the example in the repo and hardly made any changes.

[trl/trl/trainer/ppov2_trainer.py](https://github.com/huggingface/trl/blob/main/examples/scripts/ppo/ppo_tldr.py)

it takes less than 1 min for 1 training step with batch 32 on A100 GPUs

**Comment by ZxAndJb** [comment]
just my personal opinion, because the generation process does not require gradient, you can use vllm to generate the responses rather than using 'generate' method. And then concating prompts and responses for training.

**Comment by RylanSchaeffer** [comment]
I'm closing this issue because I now no longer have this issue. `.generate()` is sufficiently fast for me.

@ZxAndJb do you have code you could share for how to quickly update the policy model used by PPOv2Trainer (or some other trainer) and also used by VLLM?

**Comment by davidhughhenrymack** [comment]
I got `generate()` up to SOTA speed via the following:

- ensure model is on GPU
- load up model in 8bit

Before generation ensure no gradients and no gradient checkpointing (The latter was stealing a lot of speed), and use a batch size more appropriate than 4:
```
with torch.no_grad():
    ppo_trainer.model.gradient_checkpointing_disable()
    response_tensors = ppo_trainer.generate(
                        query_tensor=query_tensors,
                        return_prompt=False,
                        use_cache=True,
                        batch_size=64,
                    )
```

use_cache may not be necessary, just threw it in for good measure.
